lib/cache/base.ts

```diff
@@ -52,25 +63,25 @@ export class BaseCache {
         this.countersEnabled = true;
     }
 
-    stats() {
+    stats(): CacheStats {
         return {hits: this.hits, puts: this.puts, gets: this.gets};
     }
 
-    statString() {
+    statString(): string {
         const pc = this.gets ? (100 * this.hits) / this.gets : 0;
         const misses = this.gets - this.hits;
         return `${this.puts} puts; ${this.gets} gets, ${this.hits} hits, ${misses} misses (${pc.toFixed(2)}%)`;
     }
 
-    report() {
+    report(): void {
         logger.info(`${this.cacheName} ${this.details}: cache stats: ${this.statString()}`);
     }
 
-    static hash(object) {
+    static hash(object): string {
```

Comment:
Missing type, can leave unknown if we don't know/too complex to take care of right now

Reply:
we can hash anything at all that's JSON.stringifable, so `any` ? or is there a built-in that says "can be json'd" ?

Reply:
Technically not all values can be JSON.stringified (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description) but I guess any will do with a tiny comment maybe linking to this MDN link

Reply:
Right! I wondered if there was a pre-existing "thing that can be JSON'd" type (I usually make my own up in python typing)

Reply:
Elsewhere we have:
```js
function objectToHashableString(object: any): string {
    // See https://stackoverflow.com/questions/899574/which-is-best-to-use-typeof-or-instanceof/6625960#6625960
    return typeof object === 'string' ? object : JSON.stringify(object);
}```

Taking an `any` so maybe that's ok after all?

Reply:
I found another way to do this that I prefer, will check something in shortly.